### PR TITLE
MMO x DDL

### PR DESCRIPTION
**Pull request by robes**
Integration of MMO with DDL operations.

### CODE REVIEW

File: deriva/core/ermrest_model.py [hunk 1]
```diff
@@ -1,15 +1,16 @@
 
 from __future__ import annotations
 
+import base64
```
Comment (robes) [inline deriva/core/ermrest_model.py:4]: Hmmm... this must be what pycharm meant by "optimizing" imports on commit. I'm gonna not do that in the future. If this is annoying I can revert this part but its just a reordering as far as I see.

File: deriva/core/ermrest_model.py [hunk 2]
```diff
@@ -1821,23 +1823,29 @@ def create_reference(
...
 
         self.catalog.delete(self.uri_path).raise_for_status()
         del self.schema.tables[self.name]
         for fkey in self.foreign_keys:
             fkey._cleanup()
 
+        if update_mappings:
+            for fkey in self.foreign_keys:
+                mmo.prune(self.schema.model, [fkey.constraint_schema.name, fkey.constraint_name])
+            self.schema.apply()
```
Comment (karlcz) [inline deriva/core/ermrest_model.py:1847]: Doesn't this have to be self.schema.model.apply?  Changes could be global across any schema in the model, right?

Comment (robes) [reply deriva/core/ermrest_model.py]: thanks. fixed

File: deriva/core/ermrest_model.py [hunk 3]
```diff
@@ -1716,14 +1718,25 @@ def alter(
             self.schema.tables[self.name] = self
 
         if 'schema_name' in changes:
+            old_schema_name = self.schema.name
             del self.schema.tables[self.name]
             self.schema = self.schema.model.schemas[changed['schema_name']]
+            for key in self.keys:
+                if key.constraint_schema:
+                    key.constraint_schema = self.schema
```
Comment (robes) [inline deriva/core/ermrest_model.py:1731]: @karlcz I added this to (I believe) fix a bug in that the existing method wasn't updating the `key.constraint_schema`. Unless I misunderstood something... The behavior I was getting before adding this code is that that `model.apply()` was throwing a `KeyError` exception.

File: deriva/core/ermrest_model.py [hunk 4]
```diff
@@ -23,6 +24,12 @@ class NoChange (object):
 # singletone to use in APIs below
 nochange = NoChange()
 
+class UpdateMappings (str, Enum):
+    """Update Mappings flag enum"""
+    no_update = ''
+    deferred = 'deferred'
+    immediate = 'immediate'
+
```
Comment (robes) [inline deriva/core/ermrest_model.py:32]: @karlcz think this is okay this way? or would it be better to have variable instances of the enum of say `no_update = UpdateMappings.no_update` analogous to your `nochange = NoChange()` instance?

File: deriva/core/ermrest_model.py (hunk 1 above)
Comment (karlcz) [reply deriva/core/ermrest_model.py:4]: I think it's ok

File: deriva/core/ermrest_model.py (hunk 4 above)
Comment (karlcz) [reply deriva/core/ermrest_model.py:32]: I think this is fine. You can use it as `UpdateMappings.no_update` etc.

Comment (karlcz) [reply deriva/core/ermrest_model.py:32]: However, I'm not sure I like the empty string...  I think I'd also spell it out.

Comment (karlcz) [reply deriva/core/ermrest_model.py:32]: Nevermind, I see you did it to make the "falsish" test easier to write below...

File: deriva/core/ermrest_model.py (hunk 3 above)
Comment (karlcz) [reply deriva/core/ermrest_model.py:1731]: Good catch

**Comment by karlcz** [comment]
I think this is good to merge now.  I didn't actually try to run the test suite though, as I'm not sure if/how it depends on the other open PR for mmo...?

**Comment by robes** [comment]
Cool. I have run the tests. It all succeeds.

**Pull request merged**